src/clusterfuzz/_internal/bot/tasks/utasks/regression_task.py

```diff
@@ -149,114 +153,197 @@ def _testcase_reproduces_in_revision(
-        testcase, testcase_file_path, job_type, revision, output, fuzz_target)
+      In this case, `regression_task_output` is modified in the following ways:
 
-    if error:
-      # Skip this revision only on bad build errors.
-      if error.error_type == uworker_msg_pb2.REGRESSION_BAD_BUILD_ERROR:  # pylint: disable=no-member
-        continue
-      return error
+        regression_task_output.last_regression_min is set
+        regression_task_output.last_regression_max is set
 
-    if not is_crash:
-      # We've found the latest passing revision, no need to binary search.
-      output.regression_range_start = revision
-      output.regression_range_end = last_known_crashing_revision
-      return uworker_msg_pb2.Output(regression_task_output=output)  # pylint: disable=no-member
+    b. If no such revision can be found - i.e. the earliest good revision X
+      still reproduces the testcase:
 
-    last_known_crashing_revision = revision
+        None, None, output
 
-  # All most recent revisions crash.
-  return None
+      where:
 
+        output.regression_task_output = regression_task_output
+        output.regression_task_output.regression_range_start = 0
+        output.regression_task_output.regression_range_end = X
 
-def check_earliest_revisions(
-    testcase: data_types.Testcase,
-    testcase_file_path: str,
-    job_type: str,
-    revision_range: List[int],
-    fuzz_target: Optional[data_types.FuzzTarget],
-    regression_task_output: uworker_msg_pb2.RegressionTaskOutput,  # pylint: disable=no-member
-) -> Optional[uworker_msg_pb2.Output]:  # pylint: disable=no-member
-  """Check that the earliest good build does not crash.
+    c. If we timed out:
 
-  Adds information about any bad builds encountered while running to
-  `regression_task_output.build_data_list`.
+        None, None, output
 
-  Returns:
-    None if one of the earliest builds is good and does not crash.
-    An output proto if:
+      where:
+
+        output.error_type = REGRESSION_TIMED_OUT
+        output.regression_task_output = regression_task_output
+        output.regression_task_output.last_regression_max is set
+
+    d. If another error occurred:
+
+        None, None, output
+
+      where:
+
+        output.error_type indicates the error that occurred
+        output.regression_task_output = regression_task_output
+        output.regression_task_output.last_regression_max is set
 
-    a. The earliest good build crashes, in which case the regression range is
-       set to [0, min_good_revision).
-    b. An error occurred.
   """
-  # Test to see if we crash in the oldest revision we can run. This is a pre-
-  # condition for our binary search. If we do crash in that revision, it
-  # implies that we regressed between the first commit and our first revision,
-  # which we represent as 0:|min_revision|.
+  assert max_index >= 0, max_index
+  assert max_index < len(revision_list), max_index
```

Comment:
I think this could be a problem in google3/android where builds are deleted. I suppose it would be a somewhat rare race condition (if GCS deleted it as we were doing regession). WDYT?

Reply:
Right! This case is handled by the code on line 442. If the max revision we had left off at was deleted, we will choose a max revision higher than the last one we used, and in all cases confirm that the testcase reproduces at `max_index` before attempting to search for anything - see line 453.

Reply:
See also `TestFindMinRevision.test_resume_missing_crash_revision` that exercises how this function is called in that case. Do you want a corresponding test in `UTaskMainTest`?